// @framer/framer.default/code/Checkbox.tsx
import * as React from "react"
import { useState, useEffect } from "react"
import { addPropertyControls, ControlType, motion, Color } from "framer"
import { colorFromToken } from "./lib/colorFromToken"

/*
 ** CHECKBOX
 */

export interface CheckboxProps {
    tint: string
    onToggle: () => void
    onCheck: () => void
    onUncheck: () => void
    isChecked: boolean
    radius: number
    topLeft: number
    topRight: number
    bottomRight: number
    bottomLeft: number
    isMixed: boolean
    checkSize: number
    color: string
}

export function Checkbox({
    activeBackground,
    onToggle,
    onCheck,
    onUncheck,
    isChecked,
    radius,
    topLeft,
    topRight,
    bottomRight,
    bottomLeft,
    isMixed,
    checkSize,
    color,
    inactiveBackground,
}) {
    const [checked, setChecked] = useState(isChecked)
    const borderRadius = isMixed ? `${topLeft}px ${topRight}px ${bottomRight}px ${bottomLeft}px` : `${radius}px`

    useEffect(() => {
        setChecked(isChecked)
    }, [isChecked])

    return (
        <div style={containerStyle}>
            <motion.div
                style={{ ...checkboxStyle, borderRadius }}
                initial={false}
                animate={{
                    background: checked
                        ? Color(colorFromToken(activeBackground)).toValue()
                        : Color(colorFromToken(inactiveBackground)).toValue(),
                }}
                onTap={() => {
                    setChecked(!checked)
                    if (onToggle) onToggle()
                    if (onCheck && !checked) onCheck()
                    if (onUncheck && checked) onUncheck()
                }}
            >
                <svg viewBox="0 0 12 12" style={{ margin: 1, width: `${checkSize}%`, height: `${checkSize}%` }}>
                    <motion.path
                        d="M3 6l2 2 4-4"
                        fill="transparent"
                        strokeWidth="1.5"
                        stroke={color}
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeDasharray="8.5"
                        transition={{ ease: "easeOut", duration: 0.15 }}
                        initial={{
                            strokeDashoffset: checked ? 0 : 8.5,
                        }}
                        animate={{
                            strokeDashoffset: checked ? 0 : 8.5,
                        }}
                    />
                </svg>
            </motion.div>
        </div>
    )
}

Checkbox.defaultProps = {
    height: 20,
    width: 20,
    text: "Get started!",
    activeBackground: "#0099ff",
    inactiveBackground: "#F3F3F3",
    radius: 6,
    checkSize: 100,
    color: "white",
}

// Learn more: https://framer.com/api/property-controls/
addPropertyControls(Checkbox, {
    activeBackground: {
        title: "Active",
        type: ControlType.Color,
        defaultValue: Checkbox.defaultProps.activeBackground,
    },
    inactiveBackground: {
        title: "Inactive",
        type: ControlType.Color,
        defaultValue: Checkbox.defaultProps.inactiveBackground,
    },
    radius: {
        title: "Radius",
        type: ControlType.FusedNumber,
        defaultValue: Checkbox.defaultProps.radius,
        toggleKey: "isMixed",
        toggleTitles: ["Radius", "Radius per corner"],
        valueKeys: ["topLeft", "topRight", "bottomRight", "bottomLeft"],
        valueLabels: ["TL", "TR", "BR", "BL"],
        min: 0,
    },
    isChecked: {
        title: "Checked",
        type: ControlType.Boolean,
        defaultValue: true,
    },
    color: {
        title: "Check Color",
        type: ControlType.Color,
        defaultValue: Checkbox.defaultProps.color,
    },
    checkSize: {
        title: "Check Size",
        type: ControlType.Number,
        unit: "%",
        min: 0,
        max: 100,
        defaultValue: Checkbox.defaultProps.checkSize,
    },
    onToggle: { type: ControlType.EventHandler },
    onCheck: { type: ControlType.EventHandler },
    onUncheck: { type: ControlType.EventHandler },
})

const containerStyle: React.CSSProperties = {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    height: "100%",
    width: "100%",
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
}
const checkboxStyle: React.CSSProperties = {
    width: "100%",
    height: "100%",
    borderRadius: 6,
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
}
